src/tentaclio/clients/gs_client.py

```diff
@@ -0,0 +1,156 @@
+"""GS Stream client."""
+from typing import Optional, Union, Tuple, cast
+
+from google.cloud import storage
+
+from tentaclio import urls, protocols
+
+from . import base_client, decorators, exceptions
+
+
+__all__ = ["GSClient"]
+
+
+class GSClient(base_client.BaseClient["GSClient"]):
+    """GS client.
+
+    Ref: https://googleapis.dev/python/storage/latest/index.html
+    """
+
+    allowed_schemes = ["gs", "gcs"]
+
+    conn: storage.Client
+    bucket: Optional[str]
+    key_name: Optional[str]
+
+    def __init__(
+        self, url: Union[urls.URL, str]
+    ) -> None:
+        """Create a new GS client.
+
+        The client is created based on a URL of the form
+        gs://bucket/key.
+
+        Authentication and client configuration is recommended with environment variables.
+        Quoted from docs:
+            "Setting the GOOGLE_APPLICATION_CREDENTIALS and GOOGLE_CLOUD_PROJECT environment
+            variables will override the automatically configured credentials."
+
+        See documentation for more information.
+        https://googleapis.dev/python/google-api-core/latest/auth.html
+        """
+        super().__init__(url)
+
+        self.bucket = self.url.hostname or None
+        self.key_name = self.url.path[1:] if self.url.path != "" else ""
+
+        if self.bucket == "gs":
+            self.bucket = None
+
+    def _connect(self) -> storage.Client:
+        return storage.Client()
+
+    def close(self) -> None:
+        """Close the connection.
+
+        This is fake. GS doesn't allow to close gs connections.
+        """
+        # gs doesn't have close method
+        if self.closed:
+            raise ValueError("Trying to close a closed client")
+        self.closed = True
+
+    # Stream methods:
+
+    @decorators.check_conn
+    def get(
+        self, writer: protocols.ByteWriter, bucket_name: str = None, key_name: str = None
+    ) -> None:
+        """Download the contents from gs and write them in the provided writer.
+
+        Arguments:
+            :bucket_name: If not provided in the url at construction time.
+            :key_name: If not provided in the url at construction time.
+
+        Raises:
+            GSError: If a bucket or a key is not found in the given URL or args then
+            Google Cloud Exceptions: if the client raises them.
+        """
+        gs_bucket, gs_key = self._fetch_bucket_and_key(bucket_name, key_name)
+
+        self._get(writer, gs_bucket, gs_key)
```

Comment:
looking at these get/put/remove functions, you could probably just combine the hidden methods with these and then the all the tests that mock the private methods can just be tests of _fetch_bucket_and_key. i guess you get a tiny bit lower coverage wrt passing the bucket + key along to the next part of the function, and maybe its harder to test the private methods tho... so i dont really mind either way